neo/Consensus/ConsensusService.cs

```diff
@@ -225,9 +286,281 @@ private void OnPersistCompleted(Block block)
+                // Verify that there are M valid preparations, 1 Prepare Request + (M-1) Prepare responses
+                if (prepareRequest == null || prepareResponses.Count < context.M - 1) continue;
+                canRestoreView = true;
+                break;
+            }
+
+            // Only accept recovery from lower views if there were at least M valid prepare requests
+            if (message.ViewNumber < context.ViewNumber && !canRestoreView) return;
+
+            var verifiedChangeViewWitnessInvocationScripts = new byte[context.Validators.Length][];
+            if (!canRestoreView && message.ChangeViewWitnessInvocationScripts != null)
+            {
+                if (message.ChangeViewWitnessInvocationScripts.Length < context.Validators.Length) return;
+                if (message.ChangeViewTimestamps.Length < context.Validators.Length) return;
+                if (message.OriginalChangeViewNumbers.Length < context.Validators.Length) return;
+
+                var changeViewMsg = new ChangeView { NewViewNumber = message.ViewNumber };
+                int validChangeViewCount = 0;
+                for (int i = 0; i < context.Validators.Length; i++)
+                {
+                    if (message.ChangeViewWitnessInvocationScripts[i] == null) continue;
+
+                    changeViewMsg.ViewNumber = message.OriginalChangeViewNumbers[i];
+                    // Regenerate the ChangeView message
+                    var regeneratedChangeView = tempContext.RegenerateSignedPayload(changeViewMsg, (ushort) i,
+                        message.ChangeViewWitnessInvocationScripts[i], message.ChangeViewTimestamps[i]);
+                    if (!regeneratedChangeView.Verify(snap) || !PerformBasicConsensusPayloadPreChecks(regeneratedChangeView)) continue;
+                    verifiedChangeViewWitnessInvocationScripts[i] = message.ChangeViewWitnessInvocationScripts[i];
+                    validChangeViewCount++;
+                    if (validChangeViewCount < context.M) continue;
+                    canRestoreView = true;
+                    break;
+                }
+            }
+
+            if (!canRestoreView) return;
+            // We had enough valid change view messages or preparations to safely change the view number.
+            Log($"regenerating view: {message.ViewNumber}");
+
+            if (context.PrimaryIndex == context.MyIndex && prepareRequestPayload != null)
+            {
+                // If we are the primary and there was a valid prepare request payload, we will accept our own prepare
+                // request, so avoid the normal IntializeConsensus behavior.
+                context.Reset(message.ViewNumber);
+                // Since we won't want to fill the context, we will behave like a backup even though we have the primary
+                // index; we set the backup state so we can call OnPrepareRequestReceived below.
+                context.State |= ConsensusState.Backup;
```

Comment:
I think that I do not like this so much, Jeff. If the node is Primary, we should not set Backup flag to it.

Reply:
The logic is simpler if it acts as a backup, because it will never again send the prepare request. It is more complicated to make it have the Primary flag set and the logic work correctly. However, I will take a look to see how complicated it will be.

Reply:
When the node is primary but is not generating a prepare request it is operating as a backup, therefore it sets the `Backup` flag. Having the other flag set adds more code with no added benefit. One can still check if it is the primary index by checking if `context.MyIndex == context.PrimaryIndex`. Since there has less code changes required with it the way it is, I currenlty prefer leaving it the way it is.

Reply:
It would be nice if it could be changed

Reply:
Ok, i changed it so it still has the Primary flag set now :100: